src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/BulkCopy/SqlBulkCopy.Async.cs

```diff
@@ -0,0 +1,1108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Data;
+using System.Data.Common;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+
+namespace Microsoft.Data.SqlClient
+{
+    public partial class SqlBulkCopy
+    {
+        // Newly added member variables for Async modification, m = member variable to bcp.
+        private int _savedBatchSize = 0; // Save the batchsize so that changes are not affected unexpectedly.
+        private bool _hasMoreRowToCopy = false;
+
+        // we can remove this variable as it is not being used anymore.
+        private bool _isAsyncBulkCopy = false;
+        private bool _isBulkCopyingInProgress = false;
+        private SqlInternalConnectionTds.SyncAsyncLock _parserLock = null;
+
+        /// <include file='../../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlBulkCopy.xml' path='docs/members[@name="SqlBulkCopy"]/WriteToServerAsync[@name="DataRowParameter"]/*'/>
+        public async Task WriteToServerAsync(DataRow[] rows) => await WriteToServerAsync(rows, CancellationToken.None);
+
+        /// <include file='../../../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlBulkCopy.xml' path='docs/members[@name="SqlBulkCopy"]/WriteToServerAsync[@name="DataRowAndCancellationTokenParameters"]/*'/>
+        public async Task<Task> WriteToServerAsync(DataRow[] rows, CancellationToken cancellationToken)
+        {
+            Task resultTask = null;
+
+            if (rows == null)
+            {
+                throw new ArgumentNullException(nameof(rows));
+            }
+
+            if (_isBulkCopyingInProgress)
+            {
+                throw SQL.BulkLoadPendingOperation();
+            }
+
+            SqlStatistics statistics = Statistics;
+            try
+            {
+                statistics = SqlStatistics.StartTimer(Statistics);
+
+                if (rows.Length == 0)
+                {
+                    return cancellationToken.IsCancellationRequested ?
+                            Task.FromCanceled(cancellationToken) :
+                            Task.CompletedTask;
+                }
+
+                DataTable table = rows[0].Table;
+                Debug.Assert(null != table, "How can we have rows without a table?");
+                _rowStateToSkip = DataRowState.Deleted; // Don't allow deleted rows
+                _rowSource = rows;
+                _dataTableSource = table;
+                _SqlDataReaderRowSource = null;
+                _rowSourceType = ValueSourceType.RowArray;
+                _rowEnumerator = rows.GetEnumerator();
+                _isAsyncBulkCopy = true;
+                resultTask = WriteRowSourceToServerAsync(table.Columns.Count, cancellationToken); // It returns Task since _isAsyncBulkCopy = true;
```

Comment:
This block and most of the method feels a bit boiler-plate. Do some checks, setup statistics, then do the variable bit and then cleanup. Is it worth abstracting it out as a separate method do you think?

Reply:
I agree. I will look into that tomorrow.